File: src/python/topic_pubsub_py/topic_pubsub_py/sub.py
"""
Example of a Python module with a ROS 2 topic subscriber.

Roberto Masocco <robmasocco@gmail.com>
Intelligent Systems Lab <isl.torvergata@gmail.com>

May 22, 2023
"""

import sys

#! Import Python client library for ROS 2
import rclpy
from rclpy.node import Node

#! QoS profile classes and enums
import rclpy.qos

#! Our interface definition, which is a class in Python
from ros2_examples_interfaces.msg import String

#! To define a node, we need to inherit from the Node class


class Sub(rclpy.node.Node):
    """Example ROS 2 node with a subscriber to a custom topic."""

    def __init__(self) -> None:
        """
        Creates a new Sub node.
        """
        #! We need to call the base Node constructor
        super().__init__('subscriber_node')

        #! Create a custom, best-effort QoS profile
        qos_profile = rclpy.qos.QoSProfile(
            depth=1,
            history=rclpy.qos.HistoryPolicy.KEEP_LAST,
            reliability=rclpy.qos.ReliabilityPolicy.BEST_EFFORT,
            durability=rclpy.qos.DurabilityPolicy.VOLATILE)

        #! Create a subscriber to the custom topic
        #! Note that this is Python: we are defining the subscriber attribute as we
        #! create it
        self._sub = self.create_subscription(
            String,
            '/publisher_node/examples/test_topic',
            self.sub_callback,
            qos_profile)

        self.get_logger().info('Node initialized')

    #! Callback functions are methods here too, and are passed message objects
    def sub_callback(self, msg: String) -> None:
        """
        Subscriber callback function.

        :param msg: String message to parse.
        """
        self.get_logger().info('Received message: %s' % msg.data)


def main():
    #! Initialize the ROS 2 client library and context
    rclpy.init(args=sys.argv)

    #! Initialize ROS 2 node
    node = Sub()

    #! Spin the node in a basic executor
    try:
        rclpy.spin(node)
    except:
        pass
        #! If you press Ctrl-C, the handler will call rclpy.shutdown() for us
    node.destroy_node()
    # rclpy.shutdown()

    print('Subscriber terminated')


#! This is required in Python to execute a script
if __name__ == '__main__':
    main()

File: src/cpp/custom_topic_cpp/src/sub.cpp
/**
 * Subscriber definition and implementation.
 *
 * Roberto Masocco <robmasocco@gmail.com>
 *
 * November 22, 2021
 */

#include <iostream>

#include <custom_topic_cpp/sub.hpp>

/**
 * @brief Creates a Sub node.
 */
//! Call the base class constructor providing a string embedding the node name
//! Initialize other members at will
Sub::Sub()
: Node("subscriber_node")
{
  //! This time we use an explicit QoS policy by creaitng a QoS object
  //! Syntax is: rclcpp::QoS obj_name(QUEUE_DEPTH)
  //! Then, we modify the individual parameters using the setter methods and the enums
  //! For a complete reference see the qos.hpp header from rclcpp (try to open it via VS Code!)
  rclcpp::QoS topic_qos(rclcpp::KeepLast(1));
  topic_qos.best_effort(); //! Equivalenty: topic_qos.reliability(rclcpp::ReliabilityPolicy::Reliable);
  topic_qos.durability_volatile(); //! Equivalenty: topic_qos.durability(rclcpp::DurabilityPolicy::Volatile);

  //! Initialize a subscriber with create_subscription from the base class:
  //! this->create_subscription<INTERFACE_TYPE>(
  //!   TOPIC_NAME [string],
  //!   SUBSCRIPTION_QoS (...),
  //!   CALLBACK_WRAPPER (with captures and argument placeholders)
  //!   (...)
  //! );
  subscriber_ = this->create_subscription<ros2_examples_interfaces::msg::String>(
    "/publisher_node/examples/test_topic",
    topic_qos,
    std::bind(
      &Sub::msg_callback,
      this,
      std::placeholders::_1));

  //! Logging macro used to deliver a message to the logging subsystem, INFO level
  RCLCPP_INFO(this->get_logger(), "Subscriber initialized");
}

/**
 * @brief Echoes a new message.
 *
 * @param msg New message.
 */
void Sub::msg_callback(const ros2_examples_interfaces::msg::String::SharedPtr msg)
{
  //! Get the data simply by accessing the object member as the interface specifies
  RCLCPP_INFO(this->get_logger(), msg->data.c_str());
}

int main(int argc, char ** argv)
{
  //! This automatically creates the global context->DDS participant for this application
  //! and parses all ROS-specific input arguments eventually passed to the new process
  //! (and installs signal handlers)
  rclcpp::init(argc, argv);

  //! Since it's all made of shared pointers, initialize the new node as such
  //! (note: here we're calling an actual constructor but obtaining a smart pointer
  //! to the new object)
  auto sub_node = std::make_shared<Sub>();

  //! This automatically creates a default, single-threaded executor, adds to it
  //! the jobs defined by the new node, and makes the current thread tend to that
  rclcpp::spin(sub_node);

  //! Whatever happens now, does so only after rclcpp::spin has returned

  //! Shut down the global context, the DDS participant and all its
  //! endpoints, effectively terminating the connection to the DDS layer
  //! (note: this doesn't necessarily destroy the corresponding objects)
  rclcpp::shutdown();

  // Just exit
  std::cout << "Subscriber terminated" << std::endl;
  exit(EXIT_SUCCESS);
}
